### ISSUE TITLE
How to replace a custom class name for a  Yolov8 pretrained model

### ISSUE THREAD

**Issue by Jaykumaran** [ultralytics/ultralytics#4956]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

How to replace a custom class name for a  Yolov8 pretrained model for eg Person class to be replaced with names like Michael Jackson for my desired video?Without need to train custom model

**Comment by Jaykumaran** [comment]
# I found the solution to it:
# All u have to do is class mapping

Here us a sample code:
from ultralytics import YOLO

# Load the YOLO model
model = YOLO('yolov8m-pose.pt')

# Define a class mapping dictionary
class_mapping = {
    0: 'Wide Angle Seated', # The key is the class id, you may need to adjust according to your model
    # Add more mappings as needed
}

# Perform object detection on the image
results = model(source='Wide Angle Seated.jpg')

# Replace class names with custom labels in the results
for result in results:
    for cls_id, custom_label in class_mapping.items():
        if cls_id in result.names: # check if the class id is in the results
            result.names[cls_id] = custom_label # replace the class name with the custom label

# Perform object detection on the image
results = model(source='WWide Angle Seated.jpg.jpg',save=True, conf=0.7)

**Comment by glenn-jocher** [comment]
@Jaykumaran it's great to hear that you found a solution! Your approach of creating a class mapping and replacing the 'Person' class with a custom name seems like a perfectly valid way to achieve your goal. It's definitely a good idea to check if the class ID exists in the results before attempting to replace the class name. Your solution should work well for anyone wanting to display custom names for the detected classes. Thank you for sharing this with the community, and please don't hesitate to ask if you have any more questions regarding the use of YOLOv8.

**Comment by Jaykumaran** [comment]
Hi Glenn,
I'm trying to learn YoloV8/Yolov5 architecture from scratch using pytorch
right from model,dataset,loss,training, prediction.Guide me where i can
learn them.

For example,
https://youtu.be/Grir6TZbc1M?feature=shared

**Comment by glenn-jocher** [comment]
@Jaykumaran hi there,

It's great that you're enthusiastic about learning the Yolov8 architecture from scratch! 

Understanding the dataset, loss function, training, and prediction processes involved in a deep learning model like YOLOv8 is an excellent step to start with. 

To learn these components, you can refer to the original YOLO papers and other tutorial assets available online. You will find a lot of resources that explain these concepts in detail. 

Additionally, the codebase itself is a rich source of learning material. As it's implemented in PyTorch, it's quite readable and contains a lot of helpful comments. Start by taking a look at the various python files in the repository like models.py, train.py, test.py, etc to understand how the model, dataset, training, and prediction processes work.

Remember, building expertise in this field takes time. Try to implement what you learn in small projects or reproduce the results using the model.

Keep learning and experimenting!

**Comment by BitCode-in** [comment]
Hello! You can rename a class in this way.

```
import torch
from ultralytics import YOLO

model = torch.load("best.pt", map_location="cpu")

model["model"].names[0] = "New name"

torch.save(model, "save_best.pt")

model = YOLO(r"save_best.pt")
print(model.names)
```

**Comment by glenn-jocher** [comment]
@BitCode-in hi there! 👋

Yes, your approach for renaming a class in a YOLOv8 model is spot-on! Just load the model, update the class name, and re-save. Your code snippet looks great. 

Just remember that the `model.names` attribute holds all class names, so when you change the name at a specific index like `model["model"].names[0] = "New name"`, you're updating the first class. If you have multiple classes to rename, just repeat this step for other indices as needed.

Happy coding! 😄

**Comment by nabiroro** [comment]
'YOLO' object is not subscriptable

**Comment by glenn-jocher** [comment]
Hi there! It looks like you are trying to access elements of the YOLO object with subscript notation (using [ ]), which isn't directly supported. Here’s how you can correctly interact with a YOLO model:

```python
from ultralytics import YOLO

# Load your model
model = YOLO('path/to/model.pt')

# Now you can use model methods, like:
results = model.predict('path/to/image.jpg')

# Print detected class names
print(results.names)
```

Make sure to use the methods provided by the YOLO class according to the Ultralytics documentation. If you need further clarification, feel free to ask! 😊

**Comment by MlLearnerAkash** [comment]
> Hello! You can rename a class in this way.
> 
> ```
> import torch
> from ultralytics import YOLO
> 
> model = torch.load("best.pt", map_location="cpu")
> 
> model["model"].names[0] = "New name"
> 
> torch.save(model, "save_best.pt")
> 
> model = YOLO(r"save_best.pt")
> print(model.names)
> ```
While working with YOLO-V8, loading with torch, model["model"] returning `<None>`

**Comment by glenn-jocher** [comment]
You should use the `YOLO` class methods to modify class names. Here's how you can do it:

```python
from ultralytics import YOLO

# Load your model
model = YOLO('

**Comment by MlLearnerAkash** [comment]
> You should use the `YOLO` class methods to modify class names. Here's how you can do it:
> 
> ```python
> from ultralytics import YOLO
> 
> # Load your model
> model = YOLO('
> ```

That does not help because YOLO use `@property` decorator that bars any change in model property.

**Comment by glenn-jocher** [comment]
@MlLearnerAkash thank you for pointing that out. To modify class names in a YOLOv8 model, you can use the `model.names` attribute directly after loading the model. Here's how:

```python
from ultralytics import YOLO

# Load your model
model = YOLO('path/to/model.pt')

# Modify class names
model.names[0] = "New name"

# Save the updated model
model.save('path/to/save_model.pt')
```

This should allow you to update the class names without issues.

**Comment by MlLearnerAkash** [comment]
> @MlLearnerAkash thank you for pointing that out. To modify class names in a YOLOv8 model, you can use the `model.names` attribute directly after loading the model. Here's how:
> 
> ```python
> from ultralytics import YOLO
> 
> # Load your model
> model = YOLO('path/to/model.pt')
> 
> # Modify class names
> model.names[0] = "New name"
> 
> # Save the updated model
> model.save('path/to/save_model.pt')
> ```
> 
> This should allow you to update the class names without issues.

Below attaching my code snippet:
`
model = YOLO("/root/ws/train/weights/best.pt")
print("Original model names:", model.names)
class_mapping = {0: 'sponge', 1: 'obstruction', 2: 'scalpel', 3: 'incision', 4: 'woodspack', 5: 'scissors', 6: 'gauze', 7: 'snare', 8: 'black_suture', 9: 'needle', 10: 'glove', 11: 'vesiloop', 12: 'needle_holder', 13: 'forceps', 14: 'sucker', 15: 'clamp', 16: 'obstruction'}

# Modify class names
for k,v in class_mapping.items():
    model.names[k] = v


# Save the updated model
model.save('save_model.pt')
model = YOLO("save_model.pt")
print("Updated model names>>>>", model.names)
`
The output:
`
Original model names: {0: 'needle_holder', 1: 'forceps', 2: 'scalpel', 3: 'vesiloop', 4: 'incision', 5: 'needle', 6: 'scissors', 7: 'sucker', 8: 'glove', 9: 'snare', 10: 'black_suture', 11: 'gauze', 12: 'black suture', 13: 'woodspack', 14: 'sponge', 15: 'clamp', 16: 'obstruction'}
Updated model names>>>> {0: 'needle_holder', 1: 'forceps', 2: 'scalpel', 3: 'vesiloop', 4: 'incision', 5: 'needle', 6: 'scissors', 7: 'sucker', 8: 'glove', 9: 'snare', 10: 'black_suture', 11: 'gauze', 12: 'black suture', 13: 'woodspack', 14: 'sponge', 15: 'clamp', 16: 'obstruction'}
`

**Comment by glenn-jocher** [comment]
Thank you for sharing your code snippet. It seems like the changes to `model.names` are not being reflected. Please ensure you are using the latest version of the Ultralytics package. If the issue persists, consider opening a bug report on our GitHub page for further investigation.

**Comment by MlLearnerAkash** [comment]
> > @MlLearnerAkash thank you for pointing that out. To modify class names in a YOLOv8 model, you can use the `model.names` attribute directly after loading the model. Here's how:
> > ```python
> > from ultralytics import YOLO
> > 
> > # Load your model
> > model = YOLO('path/to/model.pt')
> > 
> > # Modify class names
> > model.names[0] = "New name"
> > 
> > # Save the updated model
> > model.save('path/to/save_model.pt')
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This should allow you to update the class names without issues.
> 
> Below attaching my code snippet:
> 
> ```
> model = YOLO("/root/ws/train/weights/best.pt")
> print("Original model names:", model.names)
> class_mapping = {0: 'sponge', 1: 'obstruction', 2: 'scalpel', 3: 'incision', 4: 'woodspack', 5: 'scissors', 6: 'gauze', 7: 'snare', 8: 'black_suture', 9: 'needle', 10: 'glove', 11: 'vesiloop', 12: 'needle_holder', 13: 'forceps', 14: 'sucker', 15: 'clamp', 16: 'obstruction'}
> 
> # Modify class names
> for k,v in class_mapping.items():
>     model.names[k] = v
> 
> 
> # Save the updated model
> model.save('save_model.pt')
> model = YOLO("save_model.pt")
> print("Updated model names>>>>", model.names)
> ```
> 
> The output:
> 
> ```
> Original model names: {0: 'needle_holder', 1: 'forceps', 2: 'scalpel', 3: 'vesiloop', 4: 'incision', 5: 'needle', 6: 'scissors', 7: 'sucker', 8: 'glove', 9: 'snare', 10: 'black_suture', 11: 'gauze', 12: 'black suture', 13: 'woodspack', 14: 'sponge', 15: 'clamp', 16: 'obstruction'}
> Updated model names>>>> {0: 'needle_holder', 1: 'forceps', 2: 'scalpel', 3: 'vesiloop', 4: 'incision', 5: 'needle', 6: 'scissors', 7: 'sucker', 8: 'glove', 9: 'snare', 10: 'black_suture', 11: 'gauze', 12: 'black suture', 13: 'woodspack', 14: 'sponge', 15: 'clamp', 16: 'obstruction'}
> ```
> 
> No change in model names!!!! I think there is some bug!!!

Hi @glenn-jocher, I have fixed the issue by changing the ```for k,v in class_mapping.items(): ; model.names[k] = v``` to ```model.model.names = class_mapping```

**Comment by glenn-jocher** [comment]
Thank you for sharing your solution. If the issue persists, please ensure you are using the latest version of the Ultralytics package. If it still doesn't work, consider opening a bug report on our GitHub page for further investigation.

**Comment by philipperemy** [comment]
TL;DR


```python
from ultralytics import YOLO

model = YOLO('/path/to/your/checkpoint.pt')
print(model.names)
model.model.names = {0:'new_name'} # assuming one class.
model.save('new_checkpoint.pt')

model2 = YOLO('new_checkpoint.pt')
print(model.names) # should work!
```

The trick is to use `model.model`.

**Comment by sachadee** [comment]
**The right way:**
like [philipperemy](https://github.com/philipperemy) said using `model.model`

```
from ultralytics import YOLO
model = YOLO('./weights/yolov8n.pt')
model.model.names[5] = 'car'
model.model.names[7] = 'car'
model.save('./weights/yolov8nMod.pt')
```

>  **_{0: 'person', 1: 'bicycle', 2: 'car', 3: 'motorcycle', 4: 'airplane', 5: 'car', 6: 'train', 7: 'car', 8: 'boat', 9: 'traffic light', 10: 'fire hydrant', 11: 'stop sign', 12: 'parking meter', ..._**

:metal:

**Comment by glenn-jocher** [comment]
Thanks, that’s correct—updating `model.model.names` before inference and saving with `model.save(...)` will persist your label changes; if a predictor was already initialized, reload the checkpoint or set `model.predictor.model.names = model.model.names` so results reflect the updates; see the Model.names property reference and, for open‑vocabulary models, the YOLO‑World set‑prompts guide for `set_classes(...)`. [Model.names property](https://docs.ultralytics.com/reference/engine/model/#names), [YOLO‑World set prompts](https://docs.ultralytics.com/models/yolo-world/).
